src/storage/compression/bitpacking.cpp

```diff
@@ -0,0 +1,599 @@
+	data_ptr_t width_ptr;
+
+	BitpackingState<T, PRE_CAST_TYPE> state;
+};
+
+template <class T, class PRE_CAST_TYPE>
+unique_ptr<CompressionState> BitpackingInitCompression(ColumnDataCheckpointer &checkpointer,
+                                                       unique_ptr<AnalyzeState> state) {
+	return make_unique<BitpackingCompressState<T, PRE_CAST_TYPE>>(checkpointer);
+}
+
+template <class T, class PRE_CAST_TYPE>
+void BitpackingCompress(CompressionState &state_p, Vector &scan_vector, idx_t count) {
+	auto &state = (BitpackingCompressState<T, PRE_CAST_TYPE> &)state_p;
+	VectorData vdata;
+	scan_vector.Orrify(count, vdata);
+	state.Append(vdata, count);
+}
+
+template <class T, class PRE_CAST_TYPE>
+void BitpackingFinalizeCompress(CompressionState &state_p) {
+	auto &state = (BitpackingCompressState<T, PRE_CAST_TYPE> &)state_p;
+	state.Finalize();
+}
+
+//===--------------------------------------------------------------------===//
+// Scan
+//===--------------------------------------------------------------------===//
+template <class T, class PRE_CAST_TYPE>
+struct BitpackingScanState : public SegmentScanState {
+	unique_ptr<BufferHandle> handle;
+	PhysicalType compress_type;
+
+	void (*decompress_function)(data_ptr_t, data_ptr_t, bitpacking_width_t);
+
+	explicit BitpackingScanState(ColumnSegment &segment) {
+		auto &buffer_manager = BufferManager::GetBufferManager(segment.db);
+		handle = buffer_manager.Pin(segment.block);
+
+		current_width_group_ptr =
+		    handle->node->buffer + segment.GetBlockOffset() + BitpackingPrimitives::BITPACKING_HEADER_SIZE;
+
+		type = segment.type.InternalType();
+
+		// load offset to bitpacking widths pointer
+		auto bitpacking_widths_offset = Load<idx_t>(handle->node->buffer + segment.GetBlockOffset());
+		bitpacking_width_ptr = handle->node->buffer + segment.GetBlockOffset() + bitpacking_widths_offset;
+
+		// load the bitwidth of the first vector
+		LoadCurrentBitWidth();
+	}
+
+	void LoadCurrentBitWidth() {
+		current_width = Load<bitpacking_width_t>(bitpacking_width_ptr);
+		LoadDecompressFunction();
+	}
+
+	void Skip(ColumnSegment &segment, idx_t skip_count) {
+		while (skip_count > 0) {
+			if (position_in_group + skip_count < BITPACKING_WIDTH_GROUP_SIZE) {
+				// We're not leaving this bitpacking group, we can perform all skips.
+				position_in_group += skip_count;
+				break;
+			} else {
+				// The skip crosses the current bitpacking group, we skip the remainder of this group.
+				auto skipping = BITPACKING_WIDTH_GROUP_SIZE - position_in_group;
+				position_in_group = 0;
+				current_width_group_ptr += (current_width * BITPACKING_WIDTH_GROUP_SIZE) / 8;
+
+				// Update width pointer and load new width
+				bitpacking_width_ptr -= sizeof(bitpacking_width_t);
+				LoadCurrentBitWidth();
+
+				skip_count -= skipping;
+			}
+		}
+	}
+
+	void LoadDecompressFunction() {
+		decompress_function = &BitpackingPrimitives::UnPackBlock<PRE_CAST_TYPE>;
+		return;
+	}
+
+	idx_t position_in_group = 0;
```

Comment:
Minor stylistic nitpick: we prefer to group all of the public member variables together at the top of the struct/class, and prefer not to mix functions and member variables in the same block. 

This is how we usually structure classes:

```cpp
class MyClass {
public:
    // constructors, destructors, etc
    MyClass();

   // public variables
    idx_t my_public_variable = 0;

public:
   // public functions
   void MyFunction();

private:
   // private variables
   idx_t my_private_variable;

private:
   // private functions
   void MyPrivateFunction();
};
```

Reply:
will adhere to this structure from now on!